File: providers/auth/_auth_JELLYFIN.py
# providers/auth/_auth_JELLYFIN.py
from __future__ import annotations

import secrets
from typing import Any, Mapping, MutableMapping, Optional, Dict
from urllib.parse import urljoin

try:
    from _logging import log
except Exception:
    def log(msg: str, level: str = "INFO", module: str = "AUTH", **_):
        try: print(f"[{module}] {level}: {msg}")
        except Exception: pass

from ._auth_base import AuthProvider, AuthStatus, AuthManifest

UA = "CrossWatch/1.0"
__VERSION__ = "1.3.0"
HTTP_TIMEOUT_POST = 15
HTTP_TIMEOUT_GET = 10


def _clean_base(url: str) -> str:
    u = (url or "").strip()
    if not u: return ""
    if not (u.startswith("http://") or u.startswith("https://")): u = "http://" + u
    return u if u.endswith("/") else u + "/"


def _mb_auth_value(token: Optional[str], device_id: str) -> str:
    base = f'MediaBrowser Client="CrossWatch", Device="Web", DeviceId="{device_id}", Version="1.0"'
    return f'{base}, Token="{token}"' if token else base


def _headers(token: Optional[str], device_id: str) -> Dict[str, str]:
    auth_val = _mb_auth_value(token, device_id)
    h: Dict[str, str] = {
        "Accept": "application/json",
        "User-Agent": UA,
        "Authorization": auth_val,
        "X-Emby-Authorization": auth_val,
    }
    if token: h["X-MediaBrowser-Token"] = token
    return h


def _raise_with_details(resp, default: str) -> None:
    msg = default
    try:
        j = resp.json() or {}
        msg = (j.get("ErrorMessage") or j.get("Message") or msg)
    except Exception:
        t = (getattr(resp, "text", "") or "").strip()
        if t: msg = f"{default}: {t[:200]}"
    try:
        resp.raise_for_status()
    except Exception as e:
        raise RuntimeError(msg) from e
    raise RuntimeError(msg)


class JellyfinAuth(AuthProvider):
    name = "JELLYFIN"

    def manifest(self) -> AuthManifest:
        return AuthManifest(
            name="JELLYFIN",
            label="Jellyfin",
            flow="token",
            fields=[
                {"key": "jellyfin.server",   "label": "Server URL", "type": "text",     "required": True},
                {"key": "jellyfin.username", "label": "Username",   "type": "text",     "required": True},
                {"key": "jellyfin.password", "label": "Password",   "type": "password", "required": True},
            ],
            actions={"start": True, "finish": False, "refresh": False, "disconnect": True},
            notes="Sign in with your Jellyfin account to obtain a user access token.",
        )

    def capabilities(self) -> dict:
        return {
            "features": {
                "watchlist": {"read": True, "write": True},
                "ratings":   {"read": True, "write": True},
                "watched":   {"read": True, "write": True},
                "playlists": {"read": True, "write": True},
            },
            "entity_types": ["movie", "show", "episode"],
        }

    def get_status(self, cfg: Mapping[str, Any]) -> AuthStatus:
        jf = cfg.get("jellyfin") or {}
        server = (jf.get("server") or "").strip()
        token  = (jf.get("access_token") or "").strip()
        user   = (jf.get("user") or jf.get("username") or "").strip() or None
        return AuthStatus(connected=bool(server and token), label="Jellyfin", user=user)

    def start(self, cfg: MutableMapping[str, Any], redirect_uri: str) -> Dict[str, Any]:
        import requests
        from requests import exceptions as rx

        jf = cfg.setdefault("jellyfin", {})
        base = _clean_base(jf.get("server", ""))
        user = (jf.get("username") or "").strip()
        pw   = (jf.get("password") or "").strip()
        if not base: raise RuntimeError("Malformed request: missing server")
        if not user or not pw: raise RuntimeError("Malformed request: missing username/password")

        dev_id = (jf.get("device_id") or "").strip() or secrets.token_hex(16)
        jf["device_id"] = dev_id
        jf["server"] = base

        url = urljoin(base, "Users/AuthenticateByName")
        headers = _headers(token=None, device_id=dev_id)
        headers["Content-Type"] = "application/json"
        payload = {"Username": user, "Pw": pw}

        log("Jellyfin: authenticating...", level="INFO", module="AUTH")
        try:
            r = requests.post(url, json=payload, headers=headers, timeout=HTTP_TIMEOUT_POST)
        except rx.ConnectTimeout:
            raise RuntimeError("Server not reachable: timeout")
        except rx.ReadTimeout:
            raise RuntimeError("Server not reachable: timeout")
        except rx.SSLError:
            raise RuntimeError("Server not reachable: ssl")
        except rx.ConnectionError:
            raise RuntimeError("Server not reachable: connection")
        except rx.InvalidURL:
            raise RuntimeError("Malformed request: server url")
        except rx.RequestException as e:
            raise RuntimeError(f"Server not reachable: {e.__class__.__name__}")

        if r.status_code in (401, 403): raise RuntimeError("Invalid credentials")
        if r.status_code >= 500:        raise RuntimeError(f"Server error ({r.status_code})")
        if not r.ok:                    _raise_with_details(r, "Login failed")

        data = r.json() or {}
        token = (data.get("AccessToken") or "").strip()
        if not token: raise RuntimeError("Login failed: no access token returned")

        user_obj = data.get("User") or {}
        user_id  = (user_obj.get("Id") or "").strip()
        display  = (user_obj.get("Name") or user).strip()

        try:
            me = requests.get(urljoin(base, "Users/Me"), headers=_headers(token, dev_id), timeout=HTTP_TIMEOUT_GET)
            if me.ok:
                info = me.json() or {}
                display = (info.get("Name") or display).strip()
        except Exception:
            pass

        jf["access_token"] = token
        jf["user_id"] = user_id or jf.get("user_id") or ""
        jf["user"] = display or user
        jf.pop("password", None)

        log("Jellyfin: access token stored", level="SUCCESS", module="AUTH")
        return {"ok": True, "mode": "user_token", "user_id": jf.get("user_id") or ""}

    def finish(self, cfg: MutableMapping[str, Any], **payload) -> AuthStatus:
        return self.get_status(cfg)

    def refresh(self, cfg: MutableMapping[str, Any]) -> AuthStatus:
        return self.get_status(cfg)

    def disconnect(self, cfg: MutableMapping[str, Any]) -> AuthStatus:
        jf = cfg.setdefault("jellyfin", {})
        for k in ("access_token", "user_id"):
            jf.pop(k, None)
        log("Jellyfin: disconnected", level="INFO", module="AUTH")
        return self.get_status(cfg)


PROVIDER = JellyfinAuth()
__all__ = ["PROVIDER", "JellyfinAuth", "html", "__VERSION__"]


def html() -> str:
    return r'''<div class="section" id="sec-jellyfin">
  <style>
    #sec-jellyfin .grid2{display:grid;grid-template-columns:1fr 1fr;gap:12px}
    #sec-jellyfin .inline{display:flex;gap:8px;align-items:center}
    #sec-jellyfin .sub{opacity:.7;font-size:.92em}
    #sec-jellyfin input[type="checkbox"]{transform:translateY(1px)}
    #sec-jellyfin .inp-row{display:flex;gap:12px;align-items:center}
    #sec-jellyfin .inp-row .grow{flex:1 1 auto}
    #sec-jellyfin .verify{display:flex;gap:8px;align-items:center;white-space:nowrap}
    #sec-jellyfin details.settings{margin-top:8px}
    #sec-jellyfin details.settings summary{
      position:relative;display:flex;align-items:center;gap:10px;
      padding:10px 12px;margin:-2px;border-radius:12px;cursor:pointer;list-style:none;
      background:#0b0d12;border:1px solid rgba(160,160,255,.18);
      transition:box-shadow .18s ease, border-color .18s ease;
    }
    #sec-jellyfin details.settings summary:hover{
      box-shadow:0 0 18px rgba(176,102,255,.22),0 0 18px rgba(0,209,255,.22);
      border-color:rgba(176,102,255,.35);
    }
    #sec-jellyfin details.settings summary .plex-ico{
      width:26px;height:26px;border-radius:999px;display:grid;place-items:center;
      font-weight:700;color:#fff;background:linear-gradient(180deg,#b066ff 0%,#00d1ff 100%);
      box-shadow:0 0 10px rgba(176,102,255,.6),0 0 10px rgba(0,209,255,.6)
    }
    #sec-jellyfin details.settings summary .title{font-weight:700;letter-spacing:.2px}
    #sec-jellyfin details.settings summary .hint{opacity:.75;font-size:.92em;margin-left:auto;padding-right:22px}
    #sec-jellyfin details.settings summary::after{
      content:'▸';position:absolute;right:10px;top:50%;transform:translateY(-50%);opacity:.85;color:#a9b1ff
    }
    #sec-jellyfin details.settings[open] > summary::after{transform:translateY(-50%) rotate(90deg);color:#00d1ff}
    #sec-jellyfin details.settings .wrap{margin-top:10px;display:grid;grid-template-columns:1fr 1fr;gap:16px;align-items:start}
    #sec-jellyfin .lm-head{display:grid;grid-template-columns:1fr auto auto auto auto;gap:10px;align-items:center;margin-bottom:8px}
    #sec-jellyfin .lm-head .title{font-weight:700}
    #sec-jellyfin .lm-rows{
      display:grid;gap:6px;max-height:280px;min-height:200px;
      overflow:auto;border:1px solid var(--border);border-radius:10px;padding:8px;background:#090b10
    }
    #sec-jellyfin .lm-row{display:grid;grid-template-columns:1fr 40px 40px;gap:6px;align-items:center;background:#0b0d12;border-radius:8px;padding:6px 8px}
    #sec-jellyfin .lm-row.hide{display:none}
    #sec-jellyfin .lm-name{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}
    #sec-jellyfin .lm-dot{width:16px;height:16px;border-radius:50%;border:2px solid currentColor;background:transparent;cursor:pointer;display:inline-block;vertical-align:middle}
    #sec-jellyfin .lm-dot.hist{color:#b066ff;box-shadow:0 0 6px rgba(176,102,255,.55)}
    #sec-jellyfin .lm-dot.hist.on{background:#b066ff;box-shadow:0 0 10px rgba(176,102,255,.95)}
    #sec-jellyfin .lm-dot.rate{color:#00d1ff;box-shadow:0 0 6px rgba(0,209,255,.55)}
    #sec-jellyfin .lm-dot.rate.on{background:#00d1ff;box-shadow:0 0 10px rgba(0,209,255,.95)}
    #sec-jellyfin .lm-col{display:flex;align-items:center;gap:6px}
    #sec-jellyfin .lm-filter{min-width:160px}
    #sec-jellyfin select.lm-hidden{display:none}
    #sec-jellyfin .lm-rows{scrollbar-width:thin;scrollbar-color:#b066ff #0b0d12}
    #sec-jellyfin .lm-rows::-webkit-scrollbar{width:10px}
    #sec-jellyfin .lm-rows::-webkit-scrollbar-track{background:#0b0d12;border-radius:10px}
    #sec-jellyfin .lm-rows::-webkit-scrollbar-thumb{border-radius:10px;border:2px solid #0b0d12;background:linear-gradient(180deg,#b066ff 0%,#00d1ff 100%);box-shadow:0 0 8px rgba(176,102,255,.55),0 0 8px rgba(0,209,255,.55)}
    #sec-jellyfin .inline .msg{margin-left:auto;padding:8px 12px;border-radius:12px;border:1px solid rgba(0,255,170,.18);background:rgba(0,255,170,.08);color:#b9ffd7;font-weight:600}
    #sec-jellyfin .inline .msg.warn{border-color:rgba(255,210,0,.18);background:rgba(255,210,0,.08);color:#ffe9a6}
    #sec-jellyfin .inline .msg.hidden{display:none}
  </style>

  <div class="head" onclick="toggleSection && toggleSection('sec-jellyfin')">
    <span class="chev">▶</span><strong>Jellyfin</strong>
  </div>

  <div class="body">
    <div class="grid2">
      <div>
        <label>Server URL</label>
        <div class="inp-row">
          <input id="jfy_server" class="grow" placeholder="http://host:8096/">
          <label class="verify"><input id="jfy_verify_ssl" type="checkbox"> Verify SSL</label>
        </div>
      </div>
      <div>
        <label>Username</label>
        <input id="jfy_user" placeholder="username">
      </div>
    </div>
    <div class="grid2" style="margin-top:8px">
      <div>
        <label>Password</label>
        <input id="jfy_pass" type="password" placeholder="********">
      </div>
      <div>
        <label>Access Token</label>
        <input id="jfy_tok" readonly placeholder="empty = not set">
      </div>
    </div>
    <div class="inline" style="margin-top:10px">
      <button class="btn jellyfin" onclick="try{ jfyLogin && jfyLogin(); }catch(_){;}">Sign in</button>
      <div class="muted">Username/password → user access token.</div>
      <div id="jfy_msg" class="msg hidden" role="status" aria-live="polite"></div>
    </div>
    <details class="settings">
      <summary><span class="plex-ico">⚙︎</span><span class="title">SETTINGS</span><span class="hint">Server · User · Whitelist</span></summary>
      <div class="wrap">
        <div>
          <label>Server URL</label>
          <div class="inp-row">
            <input id="jfy_server_url" class="grow" placeholder="http://host:8096/">
            <label class="verify"><input id="jfy_verify_ssl_dup" type="checkbox" onclick="(function(){var a=document.getElementById('jfy_verify_ssl'); if(a) a.checked = document.getElementById('jfy_verify_ssl_dup').checked;})();"> Verify SSL</label>
          </div>
          <div class="sub">Leave blank to discover.</div>

          <label style="margin-top:10px">Username</label>
          <input id="jfy_username" placeholder="Display name">

          <label style="margin-top:10px">User_ID</label>
          <input id="jfy_user_id" placeholder="e.g. 6f7a0b3b-... (GUID)">

          <div class="inline" style="gap:12px;margin-top:12px">
            <button class="btn" onclick="(window.jfyAuto||function(){})();">Auto-Fetch</button>
            <button class="btn" title="Load Jellyfin libraries" onclick="(window.jfyLoadLibraries||function(){})();">Load libraries</button>
            <span class="sub" style="margin-left:auto">Edit values before Save if needed.</span>
          </div>
        </div>

        <div>
          <div class="lm-head">
            <div class="title">Whitelist Libraries</div>
            <input id="jfy_lib_filter" class="lm-filter" placeholder="Filter…">
            <div class="lm-col"><span class="sub">Select all:</span></div>
            <div class="lm-col"><button id="jfy_hist_all" type="button" class="lm-dot hist" title="Toggle all History" aria-pressed="false"></button><span class="sub">History</span></div>
            <div class="lm-col"><button id="jfy_rate_all" type="button" class="lm-dot rate" title="Toggle all Ratings" aria-pressed="false"></button><span class="sub">Ratings</span></div>
          </div>
          <div id="jfy_lib_matrix" class="lm-rows"></div>
          <div class="sub" style="margin-top:6px">Empty = all libraries.</div>
          <select id="jfy_lib_history" class="lm-hidden" multiple></select>
          <select id="jfy_lib_ratings" class="lm-hidden" multiple></select>
        </div>
      </div>
    </details>
  </div>
</div>

'''

File: assets/auth/auth.jellyfin.js
// assets/auth/auth.jellyfin.js
(function () {
  "use strict";

  const Q = (s, r = document) => r.querySelector(s);
  const Qa = (s, r = document) => Array.from(r.querySelectorAll(s) || []);
  const ESC = (s) => String(s || "").replace(/[&<>"']/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));
  const SECTION = "#sec-jellyfin";
  const LIB_URL = "/api/jellyfin/libraries";

  let H = new Set();
  let R = new Set();
  let S = new Set();
  let hydrated = false;

  const put = (sel, val) => { const el = Q(sel); if (el != null) el.value = (val ?? "") + ""; };
  const maskToken = (has) => { const el = Q("#jfy_tok"); if (el) { el.value = has ? "••••••••" : ""; el.dataset.masked = has ? "1" : "0"; } };
  const visible = (el) => !!el && getComputedStyle(el).display !== "none" && !el.hidden;

  function applyFilter() {
    const qv = (Q("#jfy_lib_filter")?.value || "").toLowerCase().trim();
    Qa("#jfy_lib_matrix .lm-row").forEach((r) => {
      const name = (r.querySelector(".lm-name")?.textContent || "").toLowerCase();
      r.classList.toggle("hide", !!qv && !name.includes(qv));
    });
  }

  function syncHidden() {
    const selH = Q("#jfy_lib_history");
    const selR = Q("#jfy_lib_ratings");
    const selS = Q("#jfy_lib_scrobble");
    if (selH) selH.innerHTML = [...H].map(id => `<option selected value="${id}">${id}</option>`).join("");
    if (selR) selR.innerHTML = [...R].map(id => `<option selected value="${id}">${id}</option>`).join("");
    if (selS) selS.innerHTML = [...S].map(id => `<option selected value="${id}">${id}</option>`).join("");
  }

  function syncSelectAll() {
    const rows = Qa("#jfy_lib_matrix .lm-row:not(.hide)");
    const allHist = rows.length && rows.every(r => r.querySelector(".lm-dot.hist")?.classList.contains("on"));
    const allRate = rows.length && rows.every(r => r.querySelector(".lm-dot.rate")?.classList.contains("on"));
    const allScr = rows.length && rows.every(r => r.querySelector(".lm-dot.scr")?.classList.contains("on"));
    const h = Q("#jfy_hist_all"), r = Q("#jfy_rate_all"), s = Q("#jfy_scr_all");
    if (h) { h.classList.toggle("on", !!allHist); h.setAttribute("aria-pressed", allHist ? "true" : "false"); }
    if (r) { r.classList.toggle("on", !!allRate); r.setAttribute("aria-pressed", allRate ? "true" : "false"); }
    if (s) { s.classList.toggle("on", !!allScr); s.setAttribute("aria-pressed", allScr ? "true" : "false"); }
  }

  function renderLibraries(libs) {
    const box = Q("#jfy_lib_matrix"); if (!box) return;
    box.innerHTML = "";
    const f = document.createDocumentFragment();
    (Array.isArray(libs) ? libs : []).forEach((it) => {
      const id = String(it.key);
      const row = document.createElement("div");
      row.className = "lm-row"; row.dataset.id = id;
      row.innerHTML = `
        <div class="lm-name">${ESC(it.title)}</div>
        <button type="button" class="lm-dot hist${H.has(id) ? " on" : ""}" data-kind="history" aria-pressed="${H.has(id)}" title="Toggle History"></button>
        <button type="button" class="lm-dot rate${R.has(id) ? " on" : ""}" data-kind="ratings" aria-pressed="${R.has(id)}" title="Toggle Ratings"></button>
        <button type="button" class="lm-dot scr${S.has(id) ? " on" : ""}" data-kind="scrobble" aria-pressed="${S.has(id)}" title="Toggle Scrobble"></button>`;
      f.appendChild(row);
    });
    box.appendChild(f);
    applyFilter();
    syncHidden();
    syncSelectAll();
  }

  function repaint() {
    const libs = Qa("#jfy_lib_matrix .lm-row").map(r => ({
      key: r.dataset.id,
      title: r.querySelector(".lm-name")?.textContent || ""
    }));
    renderLibraries(libs);
  }

  async function jfyLoadLibraries() {
    try {
      const r = await fetch(LIB_URL + "?ts=" + Date.now(), { cache: "no-store" });
      const d = r.ok ? await r.json().catch(() => ({})) : {};
      const libs = Array.isArray(d?.libraries) ? d.libraries : (Array.isArray(d) ? d : []);
      renderLibraries(libs);
    } catch { renderLibraries([]); }
  }

  function jfySectionLooksEmpty() {
    const s1 = Q("#jfy_server") || Q("#jfy_server_url");
    const u1 = Q("#jfy_user") || Q("#jfy_username");
    const tok = Q("#jfy_tok");
    const vals = [s1, u1, tok].map(el => el ? String(el.value || "").trim() : "");
    return vals.every(v => !v);
  }

  async function hydrateFromConfig(force = false) {
    if (hydrated && !force) return;
    try {
      const r = await fetch("/api/config", { cache: "no-store" });
      if (!r.ok) return;
      const cfg = await r.json();
      window.__cfg = cfg;
      const jf = cfg.jellyfin || {};

      put("#jfy_server", jf.server); put("#jfy_server_url", jf.server);
      put("#jfy_user", jf.user || jf.username); put("#jfy_username", jf.user || jf.username);
      put("#jfy_user_id", jf.user_id);
      const v1 = Q("#jfy_verify_ssl"), v2 = Q("#jfy_verify_ssl_dup");
      if (v1) v1.checked = !!jf.verify_ssl;
      if (v2) v2.checked = !!jf.verify_ssl;
      maskToken(!!(jf.access_token || "").trim());

      H = new Set((jf.history?.libraries || []).map(String));
      R = new Set((jf.ratings?.libraries || []).map(String));
      S = new Set((jf.scrobble?.libraries || []).map(String));

      hydrated = true;
      await jfyLoadLibraries();
    } catch { }
  }

  function ensureHydrate() {
    const sec = Q(SECTION);
    const body = sec?.querySelector(".body");
    if (!sec || (body && !visible(body))) return;
    const force = jfySectionLooksEmpty();
    hydrateFromConfig(force);
  }

  if (!Q(SECTION)) {
    const mo = new MutationObserver(() => {
      if (Q(SECTION)) { mo.disconnect(); ensureHydrate(); }
    });
    mo.observe(document.documentElement, { childList: true, subtree: true });
  }

  document.addEventListener("click", (e) => {
    const head = Q("#sec-jellyfin .head");
    if (head && head.contains(e.target)) setTimeout(ensureHydrate, 0);
  }, true);

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", () => setTimeout(ensureHydrate, 30));
  } else {
    setTimeout(ensureHydrate, 30);
  }

  async function jfyAuto() {
    try {
      const r = await fetch("/api/jellyfin/inspect?ts=" + Date.now(), { cache: "no-store" });
      if (!r.ok) return;
      const d = await r.json();
      if (d.server_url) { put("#jfy_server", d.server_url); put("#jfy_server_url", d.server_url); }
      if (d.username)   { put("#jfy_user", d.username);     put("#jfy_username", d.username); }
      if (d.user_id)    { put("#jfy_user_id", d.user_id); }
    } catch {}
  }

  async function jfyLogin() {
    const server = (Q("#jfy_server")?.value || "").trim();
    const username = (Q("#jfy_user")?.value || "").trim();
    const password = Q("#jfy_pass")?.value || "";
    const btn = Q("button.btn.jellyfin"), msg = Q("#jfy_msg");
    if (btn) { btn.disabled = true; btn.classList.add("busy"); }
    if (msg) { msg.className = "msg hidden"; msg.textContent = ""; }
    try {
      const r = await fetch("/api/jellyfin/login", {
        method: "POST", headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ server, username, password }), cache: "no-store"
      });
      const j = await r.json().catch(() => ({}));
      if (!r.ok || j?.ok === false) { if (msg) { msg.className = "msg warn"; msg.textContent = "Login failed"; } return; }
      put("#jfy_server_url", server); put("#jfy_username", username);
      if (j?.user_id) put("#jfy_user_id", j.user_id);
      maskToken(true); if (Q("#jfy_pass")) Q("#jfy_pass").value = "";
      if (msg) { msg.className = "msg"; msg.textContent = "Jellyfin connected."; }
      await jfyLoadLibraries();
    } finally { if (btn) { btn.disabled = false; btn.classList.remove("busy"); } }
  }

  async function jfyDeleteToken() {
    const delBtn = document.querySelector('#sec-jellyfin .btn.danger');
    const msg = document.querySelector('#jfy_msg');
    if (delBtn) { delBtn.disabled = true; delBtn.classList.add('busy'); }
    if (msg) { msg.className = 'msg hidden'; msg.textContent = ''; }
    try {
      const r = await fetch('/api/jellyfin/token/delete', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: '{}',
        cache: 'no-store'
      });
      const j = await r.json().catch(() => ({}));
      if (r.ok && (j.ok !== false)) {
        const tok = document.querySelector('#jfy_tok'); if (tok) { tok.value = ''; tok.dataset.masked = '0'; }
        const pass = document.querySelector('#jfy_pass'); if (pass) pass.value = '';
        if (msg) { msg.className = 'msg'; msg.textContent = 'Access token removed.'; }
      } else {
        if (msg) { msg.className = 'msg warn'; msg.textContent = 'Could not remove token.'; }
      }
    } catch {
      if (msg) { msg.className = 'msg warn'; msg.textContent = 'Error removing token.'; }
    } finally {
      if (delBtn) { delBtn.disabled = false; delBtn.classList.remove('busy'); }
    }
  }

  function mergeJellyfinIntoCfg(cfg) {
    const v = (sel) => (Q(sel)?.value || "").trim();
    const jf = (cfg.jellyfin = cfg.jellyfin || {});
    const server = v("#jfy_server_url") || v("#jfy_server");
    const user = v("#jfy_username") || v("#jfy_user");
    if (server) jf.server = server;
    if (user) { jf.user = user; jf.username = user || jf.username || ""; }
    const uid = v("#jfy_user_id"); if (uid) jf.user_id = uid;
    const vs = Q("#jfy_verify_ssl"), vs2 = Q("#jfy_verify_ssl_dup");
    jf.verify_ssl = !!((vs && vs.checked) || (vs2 && vs2.checked));
    jf.history = Object.assign({}, jf.history || {}, { libraries: Array.from(H) });
    jf.ratings = Object.assign({}, jf.ratings || {}, { libraries: Array.from(R) });
    jf.scrobble = Object.assign({}, jf.scrobble || {}, { libraries: Array.from(S) });
    return cfg;
  }

  document.addEventListener("click", (ev) => {
    const t = ev.target; if (!(t instanceof Element)) return;
    const btn = t.closest(".lm-dot") || t.closest(".lm-col")?.querySelector(".lm-dot"); if (!btn) return;

    if (btn.id === "jfy_hist_all") {
      ev.preventDefault(); ev.stopPropagation();
      const on = !btn.classList.contains("on"); btn.classList.toggle("on", on); btn.setAttribute("aria-pressed", on ? "true" : "false");
      H = new Set();
      Qa("#jfy_lib_matrix .lm-dot.hist").forEach((b) => {
        b.classList.toggle("on", on); b.setAttribute("aria-pressed", on ? "true" : "false");
        if (on) { const r = b.closest(".lm-row"); if (r) H.add(String(r.dataset.id || "")); }
      });
      syncHidden();
      repaint();
      syncSelectAll();
      return;
    }

    if (btn.id === "jfy_rate_all") {
      ev.preventDefault(); ev.stopPropagation();
      const on = !btn.classList.contains("on"); btn.classList.toggle("on", on); btn.setAttribute("aria-pressed", on ? "true" : "false");
      R = new Set();
      Qa("#jfy_lib_matrix .lm-dot.rate").forEach((b) => {
        b.classList.toggle("on", on); b.setAttribute("aria-pressed", on ? "true" : "false");
        if (on) { const r = b.closest(".lm-row"); if (r) R.add(String(r.dataset.id || "")); }
      });
      syncHidden();
      repaint();
      syncSelectAll();
      return;
    }

    if (btn.id === "jfy_scr_all") {
      ev.preventDefault(); ev.stopPropagation();
      const on = !btn.classList.contains("on"); btn.classList.toggle("on", on); btn.setAttribute("aria-pressed", on ? "true" : "false");
      S = new Set();
      Qa("#jfy_lib_matrix .lm-dot.scr").forEach((b) => {
        b.classList.toggle("on", on); b.setAttribute("aria-pressed", on ? "true" : "false");
        if (on) { const r = b.closest(".lm-row"); if (r) S.add(String(r.dataset.id || "")); }
      });
      syncHidden();
      repaint();
      syncSelectAll();
      return;
    }

    if (btn.closest("#jfy_lib_matrix")) {
      ev.preventDefault(); ev.stopPropagation();
      const row = btn.closest(".lm-row"); if (!row) return;
      const id = String(row.dataset.id || ""), kind = btn.dataset.kind;
      const on = !btn.classList.contains("on");
      btn.classList.toggle("on", on); btn.setAttribute("aria-pressed", on ? "true" : "false");
      if (kind === "history") { (on ? H.add(id) : H.delete(id)); }
      else if (kind === "ratings") { (on ? R.add(id) : R.delete(id)); }
      else if (kind === "scrobble") { (on ? S.add(id) : S.delete(id)); }
      syncHidden();
      repaint();
      syncSelectAll();
      return;
    }
  }, true);

  document.addEventListener("input", (ev) => { if (ev.target?.id === "jfy_lib_filter") applyFilter(); }, true);

  window.jfyAuto = jfyAuto;
  window.jfyLoadLibraries = jfyLoadLibraries;
  window.mergeJellyfinIntoCfg = mergeJellyfinIntoCfg;
  window.jfyLogin = jfyLogin;
  window.jfyDeleteToken = jfyDeleteToken;

  window.registerSettingsCollector?.(mergeJellyfinIntoCfg);
  document.addEventListener("settings-collect", (e) => { try { mergeJellyfinIntoCfg(e?.detail?.cfg || (window.__cfg ||= {})); } catch {} }, true);
})();

(function(){
  const SEL = '#sec-jellyfin details.settings';
  const collapse = (root=document) => root.querySelectorAll(SEL).forEach(d=>{ d.open = false; d.removeAttribute('open'); });

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', () => collapse());
  } else {
    collapse();
  }
  new MutationObserver(muts=>{
    for (const m of muts) {
      for (const n of m.addedNodes || []) {
        if (n.nodeType === 1) collapse(n);
      }
    }
  }).observe(document.documentElement, { childList: true, subtree: true });
})();
